common/functions/src/aggregates/aggregate_retention.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::Layout;
+use std::fmt;
+use std::sync::Arc;
+
+use bytes::BytesMut;
+use common_datavalues::prelude::*;
+use common_exception::Result;
+use common_io::prelude::*;
+use serde_json::json;
+use serde_json::Value as JsonValue;
+
+use super::aggregate_function::AggregateFunction;
+use super::aggregate_function::AggregateFunctionRef;
+use super::aggregate_function_factory::AggregateFunctionDescription;
+use super::StateAddr;
+
+struct AggregateRetentionState {
+    pub events: u32,
+}
+
+impl AggregateRetentionState {
+    #[inline(always)]
+    fn add(&mut self, event: u8) {
+        self.events |= 1 << event;
+    }
+
+    fn merge(&mut self, other: &AggregateRetentionState) {
+        self.events |= other.events;
+    }
+
+    fn serialize(&self, writer: &mut BytesMut) -> Result<()> {
+        serialize_into_buf(writer, &self.events)
+    }
+
+    fn deserialize(&mut self, reader: &mut &[u8]) -> Result<()> {
+        self.events = deserialize_from_slice(reader)?;
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct AggregateRetentionFunction {
+    display_name: String,
+    events_size: u8,
+    _arguments: Vec<DataField>,
+}
+
+impl AggregateFunction for AggregateRetentionFunction {
+    fn name(&self) -> &str {
+        "AggregateRetentionFunction"
+    }
+
+    fn return_type(&self) -> Result<DataTypePtr> {
+        Ok(JsonValue::to_data_type())
+    }
+
+    fn init_state(&self, place: StateAddr) {
+        place.write(|| AggregateRetentionState { events: 0 });
+    }
+
+    fn state_layout(&self) -> std::alloc::Layout {
+        Layout::new::<AggregateRetentionState>()
+    }
+
+    fn accumulate(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        _validity: Option<&common_arrow::arrow::bitmap::Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+
+        for i in 0..input_rows {
+            for j in 0..self.events_size {
+                let column: &BooleanColumn = Series::check_get(&columns[j as usize])?;
```

Comment:
Make this `column` be prepared before the loop!

Reply:
ok!